<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Class: EmblEntry
  
    &mdash; Documentation by YARD 0.9.34
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "EmblEntry";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index (E)</a> &raquo;
    
    
    <span class="title">EmblEntry</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Class: EmblEntry
  
  
  
</h1>
<div class="box_info">
  
  <dl>
    <dt>Inherits:</dt>
    <dd>
      <span class="inheritName">Object</span>
      
        <ul class="fullTree">
          <li>Object</li>
          
            <li class="next">EmblEntry</li>
          
        </ul>
        <a href="#" class="inheritanceTree">show all</a>
      
    </dd>
  </dl>
  

  
  
  
  
  

  

  
  <dl>
    <dt>Defined in:</dt>
    <dd>LMG_Assignment3/EmblEntry.rb</dd>
  </dl>
  
</div>

<h2>Overview</h2><div class="docstring">
  <div class="discussion">
    
<h2 id="label-EmblEntry">EmblEntry</h2>

<p>Object representing each entry of the .embl file (gene’s sequence informatiion)</p>

<h2 id="label-Summary">Summary</h2>

<p>Class used to represent information about the sequences of a given gene or list of genes, employing objects and methods of BioRuby gem</p>


  </div>
</div>
<div class="tags">
  

</div>



  <h2>Instance Attribute Summary <small><a href="#" class="summary_toggle">collapse</a></small></h2>
  <ul class="summary">
    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#accession-instance_method" title="#accession (instance method)">#<strong>accession</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set method for the accession ID of a gene/entry return [String] accession ID.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#chromosomal_coordinates-instance_method" title="#chromosomal_coordinates (instance method)">#<strong>chromosomal_coordinates</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set method for the chromosomal coordinates of a given gene return [Array&lt;Integer&gt;] the chromosomal coordinates as a two-elemnt array.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#entry_sequence-instance_method" title="#entry_sequence (instance method)">#<strong>entry_sequence</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set method for the genomic sequence of a given gene return [Bio::Sequence] the genomic sequence.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#gene_locus-instance_method" title="#gene_locus (instance method)">#<strong>gene_locus</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set method for the gene locus name of a given gene return [Array&lt;Integer&gt;] the gene locus name.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#seq_id-instance_method" title="#seq_id (instance method)">#<strong>seq_id</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set method for the sequence ID of a gene/entry return [Array&lt;Integer&gt;] the sequence ID.</p>
</div></span>
  
</li>

    
      <li class="public ">
  <span class="summary_signature">
    
      <a href="#source-instance_method" title="#source (instance method)">#<strong>source</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
    
    
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Get/Set method for the source of the annotation for a given gene return [String] source of annotation.</p>
</div></span>
  
</li>

    
  </ul>




  
    <h2>
      Instance Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#add_cttctt_repeat-instance_method" title="#add_cttctt_repeat (instance method)">#<strong>add_cttctt_repeat</strong>(coordinates, start_motif, end_motif, strand)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Adds a found CTTCTT motif as feature of the EMBL Bio::Sequence object.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#annotate_source_gene-instance_method" title="#annotate_source_gene (instance method)">#<strong>annotate_source_gene</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>It annotates the gene locus of the gene as a feature of the Bio::Sequence (genomic sequence) object return [void].</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#convert_to_chromosomal_coordinates-instance_method" title="#convert_to_chromosomal_coordinates (instance method)">#<strong>convert_to_chromosomal_coordinates</strong>(start_motif, end_motif)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Converts a motif gene’s sequence coordenates to chromosomal coordenates.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#initialize-instance_method" title="#initialize (instance method)">#<strong>initialize</strong>(entry, start_entry, end_entry)  &#x21d2; EmblEntry </a>
    

    
  </span>
  
  
    <span class="note title constructor">constructor</span>
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>It creates a new instance of EmblEntry.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#no_params_method-instance_method" title="#no_params_method (instance method)">#<strong>no_params_method</strong>  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>It annotates the gene locus of the gene as a feature of the Bio::Sequence (genomic sequence) object return [void].</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#process_cttctt_repeats-instance_method" title="#process_cttctt_repeats (instance method)">#<strong>process_cttctt_repeats</strong>(coordinates_to_use, regex_positive = REGEX_POSITIVE, regex_complementary = REGEX_COMPLEMENT)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Finds CTTCTT repeats in exons in either the positive strand or the complement.</p>
</div></span>
  
</li>

      
    </ul>
  

<div id="constructor_details" class="method_details_list">
  <h2>Constructor Details</h2>
  
    <div class="method_details first">
  <h3 class="signature first" id="initialize-instance_method">
  
    #<strong>initialize</strong>(entry, start_entry, end_entry)  &#x21d2; <tt><span class='object_link'><a href="" title="EmblEntry (class)">EmblEntry</a></span></tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>It creates a new instance of EmblEntry</p>

<p>return [EmblEntry] Just created EMBL entry</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>entry</span>
      
      
        <span class='type'>(<tt>Bio::EMBL</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the BioRuby EMBL entry of a given gene</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>start_entry</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the start chromosomal coordinate of the gene</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>end_entry</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the en chromosomal coordinate of the gene</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


49
50
51
52
53
54
55
56</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 49</span>

<span class='kw'>def</span> <span class='id identifier rubyid_initialize'>initialize</span><span class='lparen'>(</span><span class='id identifier rubyid_entry'>entry</span><span class='comma'>,</span> <span class='id identifier rubyid_start_entry'>start_entry</span><span class='comma'>,</span> <span class='id identifier rubyid_end_entry'>end_entry</span><span class='rparen'>)</span>
    <span class='ivar'>@accession</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_accession'>accession</span>  <span class='comment'># get accesion ID from entry
</span>    <span class='ivar'>@seq_id</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_entry_id'>entry_id</span>
    <span class='ivar'>@entry_sequence</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>  <span class='comment'># get nucleotide sequence of the gene
</span>    <span class='ivar'>@chromosomal_coordinates</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_start_entry'>start_entry</span><span class='comma'>,</span> <span class='id identifier rubyid_end_entry'>end_entry</span><span class='rbracket'>]</span> <span class='comment'># create an array with the chromosomic coordinates of the gene
</span>    <span class='ivar'>@source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span>
    <span class='ivar'>@gene_locus</span> <span class='op'>=</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
  
</div>

  <div id="instance_attr_details" class="attr_details">
    <h2>Instance Attribute Details</h2>
    
      
      <span id="accession=-instance_method"></span>
      <div class="method_details first">
  <h3 class="signature first" id="accession-instance_method">
  
    #<strong>accession</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set method for the accession ID of a gene/entry return [String] accession ID</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


31
32
33</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 31</span>

<span class='kw'>def</span> <span class='id identifier rubyid_accession'>accession</span>
  <span class='ivar'>@accession</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="chromosomal_coordinates=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="chromosomal_coordinates-instance_method">
  
    #<strong>chromosomal_coordinates</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set method for the chromosomal coordinates of a given gene return [Array&lt;Integer&gt;] the chromosomal coordinates as a two-elemnt array</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


21
22
23</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 21</span>

<span class='kw'>def</span> <span class='id identifier rubyid_chromosomal_coordinates'>chromosomal_coordinates</span>
  <span class='ivar'>@chromosomal_coordinates</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="entry_sequence=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="entry_sequence-instance_method">
  
    #<strong>entry_sequence</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set method for the genomic sequence of a given gene return [Bio::Sequence] the genomic sequence</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


16
17
18</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 16</span>

<span class='kw'>def</span> <span class='id identifier rubyid_entry_sequence'>entry_sequence</span>
  <span class='ivar'>@entry_sequence</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="gene_locus=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="gene_locus-instance_method">
  
    #<strong>gene_locus</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set method for the gene locus name of a given gene return [Array&lt;Integer&gt;] the gene locus name</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


41
42
43</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 41</span>

<span class='kw'>def</span> <span class='id identifier rubyid_gene_locus'>gene_locus</span>
  <span class='ivar'>@gene_locus</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="seq_id=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="seq_id-instance_method">
  
    #<strong>seq_id</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set method for the sequence ID of a gene/entry return [Array&lt;Integer&gt;] the sequence ID</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


36
37
38</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 36</span>

<span class='kw'>def</span> <span class='id identifier rubyid_seq_id'>seq_id</span>
  <span class='ivar'>@seq_id</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      
      <span id="source=-instance_method"></span>
      <div class="method_details ">
  <h3 class="signature " id="source-instance_method">
  
    #<strong>source</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Get/Set method for the source of the annotation for a given gene return [String] source of annotation</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


26
27
28</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 26</span>

<span class='kw'>def</span> <span class='id identifier rubyid_source'>source</span>
  <span class='ivar'>@source</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>


  <div id="instance_method_details" class="method_details_list">
    <h2>Instance Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="add_cttctt_repeat-instance_method">
  
    #<strong>add_cttctt_repeat</strong>(coordinates, start_motif, end_motif, strand)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Adds a found CTTCTT motif as feature of the EMBL Bio::Sequence object</p>

<p>return [void]</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>coordinates</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the start..end coordinates of the motif (in the gene’s sequence or chromosome)</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>start_motif</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the start coordinate of the motif (gene’s sequence or chromosomal)</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>end_motif</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the end coordinate of the motif (gene’s sequence or chromosomal)</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>strand</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>genomic strand were motif can be found (+ or -)</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


89
90
91
92
93
94
95
96
97
98</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 89</span>

<span class='kw'>def</span> <span class='id identifier rubyid_add_cttctt_repeat'>add_cttctt_repeat</span><span class='lparen'>(</span><span class='id identifier rubyid_coordinates'>coordinates</span><span class='comma'>,</span> <span class='id identifier rubyid_start_motif'>start_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span>
    <span class='kw'>return</span> <span class='kw'>if</span> <span class='ivar'>@entry_sequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_any?'>any?</span> <span class='lbrace'>{</span> <span class='op'>|</span><span class='id identifier rubyid_f'>f</span><span class='op'>|</span> <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>cttctt_repeat</span><span class='tstring_end'>&#39;</span></span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span> <span class='op'>==</span> <span class='id identifier rubyid_coordinates'>coordinates</span> <span class='rbrace'>}</span>   <span class='comment'># check if this motif has already been annotated
</span>
    <span class='id identifier rubyid_f1'>f1</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>cttctt_repeat</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_coordinates'>coordinates</span><span class='rparen'>)</span> <span class='comment'># New feature of the Embl biosequence object --&gt; repetitive motifs &quot;CTTCTT&quot;
</span>    <span class='id identifier rubyid_f1'>f1</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>start</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_start_motif'>start_motif</span><span class='rparen'>)</span><span class='rparen'>)</span>
    <span class='id identifier rubyid_f1'>f1</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>end</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span><span class='rparen'>)</span><span class='rparen'>)</span>
    <span class='id identifier rubyid_f1'>f1</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span><span class='rparen'>)</span>
    <span class='id identifier rubyid_f1'>f1</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>SO_Name</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>repeat_region</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
    <span class='ivar'>@entry_sequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_f1'>f1</span>  <span class='comment'># record as feature
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="annotate_source_gene-instance_method">
  
    #<strong>annotate_source_gene</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>It annotates the gene locus of the gene as a feature of the Bio::Sequence (genomic sequence) object return [void]</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 61</span>

<span class='kw'>def</span> <span class='id identifier rubyid_annotate_source_gene'>annotate_source_gene</span>
    <span class='ivar'>@entry_sequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span>
      <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span>
      <span class='id identifier rubyid_qual'>qual</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span>
      
      <span class='comment'># Annotate source
</span>      <span class='kw'>if</span> <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>source</span><span class='tstring_end'>&#39;</span></span>
        <span class='ivar'>@source</span> <span class='op'>=</span> <span class='id identifier rubyid_qual'>qual</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>db_xref</span><span class='tstring_end'>&#39;</span></span><span class='rbracket'>]</span>
      <span class='kw'>end</span>
      
      <span class='comment'># Annotate gene locus for the sequence
</span>      <span class='kw'>if</span> <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&#39;</span></span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_position'>position</span><span class='period'>.</span><span class='id identifier rubyid_match'>match</span><span class='lparen'>(</span><span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>^(complement\()?(\d+\.\.\d+)(\))?$</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span> <span class='op'>&amp;&amp;</span> <span class='ivar'>@gene_locus</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span>
        <span class='ivar'>@gene_locus</span> <span class='op'>=</span> <span class='id identifier rubyid_qual'>qual</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&#39;</span></span><span class='rbracket'>]</span>
      <span class='kw'>end</span>

    <span class='kw'>end</span>
    <span class='kw'>return</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="convert_to_chromosomal_coordinates-instance_method">
  
    #<strong>convert_to_chromosomal_coordinates</strong>(start_motif, end_motif)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Converts a motif gene’s sequence coordenates to chromosomal coordenates</p>

<p>return [Array&lt;Integer&gt;] the chromosomal coordinates of the CTTCTT motif</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>start_motif</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the start coordinate of the motif (gene’s sequence or chromosomal)</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>end_motif</span>
      
      
        <span class='type'>(<tt>Integer</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the end coordinate of the motif (gene’s sequence or chromosomal)</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


105
106
107
108
109
110
111</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 105</span>

<span class='kw'>def</span> <span class='id identifier rubyid_convert_to_chromosomal_coordinates'>convert_to_chromosomal_coordinates</span><span class='lparen'>(</span><span class='id identifier rubyid_start_motif'>start_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span><span class='rparen'>)</span>
    <span class='id identifier rubyid_start_chr'>start_chr</span> <span class='op'>=</span> <span class='ivar'>@chromosomal_coordinates</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span>
    <span class='id identifier rubyid_length_motif'>length_motif</span> <span class='op'>=</span> <span class='id identifier rubyid_end_motif'>end_motif</span> <span class='op'>-</span> <span class='id identifier rubyid_start_motif'>start_motif</span>   <span class='comment'># length of the motif
</span>    <span class='id identifier rubyid_start_motif'>start_motif</span> <span class='op'>=</span> <span class='id identifier rubyid_start_motif'>start_motif</span> <span class='op'>+</span> <span class='id identifier rubyid_start_chr'>start_chr</span> <span class='op'>-</span> <span class='int'>1</span> <span class='comment'># - 1 so 1 + 1 is still 1
</span>    <span class='id identifier rubyid_end_motif'>end_motif</span> <span class='op'>=</span> <span class='id identifier rubyid_start_motif'>start_motif</span> <span class='op'>+</span> <span class='id identifier rubyid_length_motif'>length_motif</span>   <span class='comment'># end coordinate of the motif in the chromosome
</span>    <span class='kw'>return</span> <span class='lbracket'>[</span><span class='id identifier rubyid_start_motif'>start_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span><span class='rbracket'>]</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="no_params_method-instance_method">
  
    #<strong>no_params_method</strong>  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>It annotates the gene locus of the gene as a feature of the Bio::Sequence (genomic sequence) object return [void]</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 61</span>

<span class='kw'>def</span> <span class='id identifier rubyid_annotate_source_gene'>annotate_source_gene</span>
    <span class='ivar'>@entry_sequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span>
      <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span>
      <span class='id identifier rubyid_qual'>qual</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span>
      
      <span class='comment'># Annotate source
</span>      <span class='kw'>if</span> <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>source</span><span class='tstring_end'>&#39;</span></span>
        <span class='ivar'>@source</span> <span class='op'>=</span> <span class='id identifier rubyid_qual'>qual</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>db_xref</span><span class='tstring_end'>&#39;</span></span><span class='rbracket'>]</span>
      <span class='kw'>end</span>
      
      <span class='comment'># Annotate gene locus for the sequence
</span>      <span class='kw'>if</span> <span class='id identifier rubyid_featuretype'>featuretype</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&#39;</span></span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_position'>position</span><span class='period'>.</span><span class='id identifier rubyid_match'>match</span><span class='lparen'>(</span><span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>^(complement\()?(\d+\.\.\d+)(\))?$</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span> <span class='op'>&amp;&amp;</span> <span class='ivar'>@gene_locus</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span>
        <span class='ivar'>@gene_locus</span> <span class='op'>=</span> <span class='id identifier rubyid_qual'>qual</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>gene</span><span class='tstring_end'>&#39;</span></span><span class='rbracket'>]</span>
      <span class='kw'>end</span>

    <span class='kw'>end</span>
    <span class='kw'>return</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="process_cttctt_repeats-instance_method">
  
    #<strong>process_cttctt_repeats</strong>(coordinates_to_use, regex_positive = REGEX_POSITIVE, regex_complementary = REGEX_COMPLEMENT)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Finds CTTCTT repeats in exons in either the positive strand or the complement. It retrieves coordinates dependent of the reference system (chromosome or sequence)</p>

<p>return [void]</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>coordinates_to_use</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>reference system to calculate motif’s coordinates</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>regex_positive</span>
      
      
        <span class='type'>(<tt>Regexp</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>REGEX_POSITIVE</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>regular expression for searching for the CTTCTT pattern in the positive strand</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>regex_complementary</span>
      
      
        <span class='type'>(<tt>Regeexp</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>REGEX_COMPLEMENT</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>regular expresion for searching for the CTTCTT pattern in the complementary strand</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'LMG_Assignment3/EmblEntry.rb', line 119</span>

<span class='kw'>def</span> <span class='id identifier rubyid_process_cttctt_repeats'>process_cttctt_repeats</span><span class='lparen'>(</span><span class='id identifier rubyid_coordinates_to_use'>coordinates_to_use</span><span class='comma'>,</span> <span class='id identifier rubyid_regex_positive'>regex_positive</span><span class='op'>=</span><span class='const'>REGEX_POSITIVE</span><span class='comma'>,</span> <span class='id identifier rubyid_regex_complementary'>regex_complementary</span> <span class='op'>=</span> <span class='const'><span class='object_link'><a href="top-level-namespace.html#REGEX_COMPLEMENT-constant" title="REGEX_COMPLEMENT (constant)">REGEX_COMPLEMENT</a></span></span><span class='rparen'>)</span>

  <span class='ivar'>@entry_sequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
    <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>exon</span><span class='tstring_end'>&#39;</span></span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span><span class='period'>.</span><span class='id identifier rubyid_match?'>match?</span><span class='lparen'>(</span><span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>^(complement\()?(\d+\.\.\d+)(\))?$</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span>    <span class='comment'># only search in exons
</span>    
    <span class='id identifier rubyid_start_exon'>start_exon</span><span class='comma'>,</span> <span class='id identifier rubyid_end_exon'>end_exon</span> <span class='op'>=</span> <span class='id identifier rubyid_get_coordinates'><span class='object_link'><a href="top-level-namespace.html#get_coordinates-instance_method" title="#get_coordinates (method)">get_coordinates</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span><span class='rparen'>)</span>  <span class='comment'># retrive exonic coordenates
</span>    <span class='id identifier rubyid_exon_sequence'>exon_sequence</span> <span class='op'>=</span> <span class='ivar'>@entry_sequence</span><span class='period'>.</span><span class='id identifier rubyid_subseq'>subseq</span><span class='lparen'>(</span><span class='id identifier rubyid_start_exon'>start_exon</span><span class='comma'>,</span> <span class='id identifier rubyid_end_exon'>end_exon</span><span class='rparen'>)</span>  <span class='comment'># get exon subsequence from exon coordinates
</span>    
    <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_exon_sequence'>exon_sequence</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span> <span class='op'>||</span> <span class='id identifier rubyid_exon_sequence'>exon_sequence</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span>    <span class='comment'># control: is empty?
</span>    
    <span class='kw'>if</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_position'>position</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>complement</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>    <span class='comment'># get directionality of exon (positive or complement strand)
</span>      <span class='id identifier rubyid_regex'>regex</span> <span class='op'>=</span> <span class='id identifier rubyid_regex_complementary'>regex_complementary</span>
      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>-</span><span class='tstring_end'>&#39;</span></span>
      <span class='id identifier rubyid_coordinates_format'>coordinates_format</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>complement(%s)</span><span class='tstring_end'>&quot;</span></span> <span class='comment'># specify the format in which the coordenates are given 
</span>    <span class='kw'>else</span>
      <span class='id identifier rubyid_regex'>regex</span> <span class='op'>=</span> <span class='id identifier rubyid_regex_positive'>regex_positive</span>
      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>+</span><span class='tstring_end'>&#39;</span></span>
      <span class='id identifier rubyid_coordinates_format'>coordinates_format</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>%s</span><span class='tstring_end'>&quot;</span></span>
    <span class='kw'>end</span>
  
    
    <span class='id identifier rubyid_matches_motifs'>matches_motifs</span> <span class='op'>=</span> <span class='id identifier rubyid_find_motifs'><span class='object_link'><a href="top-level-namespace.html#find_motifs-instance_method" title="#find_motifs (method)">find_motifs</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_exon_sequence'>exon_sequence</span><span class='comma'>,</span> <span class='id identifier rubyid_regex'>regex</span><span class='rparen'>)</span>  <span class='comment'># find motifs and calculate its coordenates if any (exon is the reference system). Works for multiple motifs
</span>    <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_matches_motifs'>matches_motifs</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>||</span> <span class='id identifier rubyid_matches_motifs'>matches_motifs</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span>
  
    <span class='comment'># For each matched motif in exon sequence
</span>    <span class='id identifier rubyid_matches_motifs'>matches_motifs</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_match'>match</span><span class='op'>|</span>
      <span class='id identifier rubyid_start_match'>start_match</span> <span class='op'>=</span> <span class='id identifier rubyid_match'>match</span><span class='lbracket'>[</span><span class='symbol'>:start</span><span class='rbracket'>]</span> <span class='comment'># get exonic coordenates of motif
</span>      <span class='id identifier rubyid_end_match'>end_match</span> <span class='op'>=</span> <span class='id identifier rubyid_match'>match</span><span class='lbracket'>[</span><span class='symbol'>:end</span><span class='rbracket'>]</span>
      <span class='comment'># Conversion from exon positions to entry sequence positions
</span>      <span class='id identifier rubyid_start_motif'>start_motif</span> <span class='op'>=</span> <span class='id identifier rubyid_start_exon'>start_exon</span> <span class='op'>+</span> <span class='id identifier rubyid_start_match'>start_match</span>  <span class='comment'># +1 for object index in ruby, -1 because of the sum of coordenates
</span>      <span class='id identifier rubyid_end_motif'>end_motif</span> <span class='op'>=</span> <span class='id identifier rubyid_start_exon'>start_exon</span> <span class='op'>+</span> <span class='id identifier rubyid_end_match'>end_match</span> <span class='op'>-</span> <span class='int'>1</span>  <span class='comment'># -1 because of the sum of coodenates
</span>    
       <span class='comment'># If chromosomal coordinates, it recalculates those motif coordinates in the chromosomal coordinates
</span>      <span class='kw'>if</span> <span class='id identifier rubyid_coordinates_to_use'>coordinates_to_use</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>chromosomal coordinates</span><span class='tstring_end'>&#39;</span></span>
          <span class='id identifier rubyid_start_motif'>start_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span> <span class='op'>=</span> <span class='id identifier rubyid_convert_to_chromosomal_coordinates'>convert_to_chromosomal_coordinates</span><span class='lparen'>(</span><span class='id identifier rubyid_start_motif'>start_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span><span class='rparen'>)</span>
      <span class='kw'>end</span>

      <span class='id identifier rubyid_coordinates'>coordinates</span> <span class='op'>=</span> <span class='id identifier rubyid_coordinates_format'>coordinates_format</span> <span class='op'>%</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_motif'>start_motif</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_motif'>end_motif</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_add_cttctt_repeat'>add_cttctt_repeat</span><span class='lparen'>(</span><span class='id identifier rubyid_coordinates'>coordinates</span><span class='comma'>,</span> <span class='id identifier rubyid_start_motif'>start_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_end_motif'>end_motif</span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span> <span class='comment'># add motif as feature of the Bio::Sequence object
</span>      
    <span class='kw'>end</span>
  <span class='kw'>end</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Wed Dec  6 07:05:08 2023 by
  <a href="https://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.34 (ruby-3.0.0).
</div>

    </div>
  </body>
</html>