{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "false"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#----------------- LOAD EXTERNAL LIBRARIES AND MODULES-------------------------------------------------------------------\n",
    "require 'bio'\n",
    "require 'net/http'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#------------------------------ INPUT FILES NAMES AS ARGUMENTS FROM  COMMAND LINE ---------------------------------------\n",
    "\n",
    "# Check for one common error: not specifying the correct number of files needed for the program to run\n",
    "if ARGV.length != 1\n",
    "    abort \"Incorrect number of files passed. Two files names must be specified: input list of genes and name for final_report\"\n",
    "end\n",
    "  \n",
    "# Check for second common error: incorrect usage, files in incorrect order or wrong name passed.\n",
    "#if ARGV[0] == \"ArabidopsisSubNetwork_GeneList.txt\"\n",
    "if ARGV[0] == 'soloungen.txt'\n",
    "        gene_file = ARGV[0]\n",
    "else \n",
    "    abort \"Incorrect order of files passed. Usage: ruby main.rb ArabidopsisSubNetwork_GeneList.txt\"\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":get_coordinates"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------ PUBLIC INSTANCE METHODS ---------------------------------------\n",
    "\n",
    "# Get response from URL\n",
    "def fetch(uri_str) \n",
    "    address = URI(uri_str)  \n",
    "    response = Net::HTTP.get_response(address)\n",
    "    case response\n",
    "      when Net::HTTPSuccess then\n",
    "        return response.body\n",
    "      else\n",
    "        raise Exception, \"Something went wrong... the call to #{uri_str} failed; type #{response.class}\"\n",
    "        response = false\n",
    "        return response\n",
    "    end\n",
    "end\n",
    "\n",
    "# Get response with NCBI E-Utils, given a database, file format and gene locus\n",
    "def ncbi_fetch(database, file_format, id)\n",
    "    url = \"http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=#{database}&format=#{file_format}&id=#{id}\"\n",
    "    #puts url\n",
    "    res = fetch(url)\n",
    "    return res\n",
    "end\n",
    "\n",
    "\n",
    "# Transform locus names file into string of coma separated locus names, while checking A. thaliana format\n",
    "def read_from_file(filename)\n",
    "    gene_list = []\n",
    "    gene_file = File.open(filename, 'r')\n",
    "    gene_file.readlines.each do |line|  \n",
    "        locus_name=line.chomp\n",
    "        if locus_name !~ /A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/ # Check for correct format of gene locus name\n",
    "            abort \"Locus name #{locus_name} does not meet the correct format. Please define locus names as AT0g00000\"\n",
    "        end\n",
    "        gene_list << locus_name \n",
    "    end\n",
    "    gene_file.close\n",
    "    gene_ids = gene_list.join(\",\")\n",
    "    return gene_ids\n",
    "end\n",
    "\n",
    "# Gets coordinates from string\n",
    "def get_coordinates(coordinates_string)\n",
    "    position = coordinates_string.match(/\\d+\\.\\.\\d+/)[0]\n",
    "    start_coordinate = position.split('..')[0].to_i\n",
    "    end_coordinate = position.split('..')[1].to_i\n",
    "    return start_coordinate, end_coordinate\n",
    "end\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#-------------------------------------------MAIN CODE ------------------------------------------------------------------\n",
    "# BEFORE EVERYTHING, SEARCHING EMBL FILE (Task 1)\n",
    "\n",
    "# 1:  Using BioRuby, examine the sequences of the ~167 Arabidopsis genes from the last assignment by retrieving them from whatever database you wish #\n",
    "# ArabidopsisSubNetwork file -> each gene locus (lines) -> search for embl file -> add to 'AT_sequences.embl'\n",
    "\n",
    "puts \"Processing #{gene_file} file, this might take a while...\"\n",
    "\n",
    "# Create string with gene IDs from file\n",
    "gene_ids = read_from_file(gene_file)\n",
    "\n",
    "        # Only one search with list of IDs\n",
    "response_body = ncbi_fetch(database = 'ensemblgenomesgene',file_format = 'embl', id = gene_ids)\n",
    "\n",
    "output_file = File.open('AT_sequences.embl', 'w')\n",
    "output_file.write(response_body)\n",
    "output_file.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":scan_repetitive_features"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class EmblProcessor\n",
    "    \n",
    "    attr_accessor :entries\n",
    "\n",
    "    # the emblprocessor is the embl flat file so entries are instances of the class entry\n",
    "    def initialize(filename)\n",
    "        @entries = scan_repetitive_features(filename)\n",
    "    end\n",
    "\n",
    "    def scan_repetitive_features(filename)\n",
    "        embl_file = Bio::FlatFile.auto(filename)    # crea el objeto embl flat file\n",
    "        entries = []\n",
    "\n",
    "        embl_file.each_entry do |entry| # como antes recorre las entrys\n",
    "            entries << entry   # add each entry instance to the array with all the entrys\n",
    "        end\n",
    "        \n",
    "        return entries\n",
    "    end  \n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":initialize"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class EmblEntry\n",
    "\n",
    "    attr_accessor :entry_sequence, :chromosomal_coordinates, :source\n",
    "\n",
    "    def initialize\n",
    "        @sequence_entry = entry.to_biosequence\n",
    "        @chromosomal_coordinates = [start_entry, end_entry]\n",
    "        @source = \"\"\n",
    "    end\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class EmblProcessor\n",
    "  attr_accessor :entries\n",
    "\n",
    "  def initialize(filename)\n",
    "    @entries = scan_repetitive_features(filename)\n",
    "  end\n",
    "\n",
    "  def scan_repetitive_features(filename)\n",
    "    embl_file = Bio::FlatFile.auto(filename)\n",
    "    entries = []\n",
    "\n",
    "    embl_file.each_entry do |entry|\n",
    "      start_entry, end_entry = get_coordinates(entry.definition)\n",
    "      embl_entry = EmblEntry.new(entry, start_entry, end_entry)\n",
    "      embl_entry.process_features\n",
    "      entries << embl_entry\n",
    "    end\n",
    "\n",
    "    entries\n",
    "  end\n",
    "\n",
    "  def load_to_gff(coordinates_to_use)\n",
    "    gff = Bio::GFF::GFF3.new\n",
    "    count = 1\n",
    "\n",
    "    @entries.each do |embl_entry|\n",
    "      embl_entry.process_cttctt_repeats(coordinates_to_use)\n",
    "\n",
    "      embl_entry.entry_sequence.features.each do |feature|\n",
    "        next unless feature.feature == 'cttctt_repeat'\n",
    "\n",
    "        qual = feature.assoc\n",
    "        attributes = [{ 'ID' => \"repeat_region_#{count}\", 'Name' => 'CTTCTT_motif' }]\n",
    "\n",
    "        seqid = embl_entry.seq_id\n",
    "\n",
    "        gff.records << Bio::GFF::GFF3::Record.new(\n",
    "          seqid,            # seqID\n",
    "          qual['source'],   # source\n",
    "          qual['SO_Name'],  # feature type\n",
    "          qual['start'],    # start\n",
    "          qual['end'],      # end\n",
    "          nil,              # score\n",
    "          qual['strand'],   # strand\n",
    "          nil,              # phase\n",
    "          attributes[0]     # attributes\n",
    "        )\n",
    "        count += 1\n",
    "      end\n",
    "    end\n",
    "\n",
    "    gff\n",
    "  end\n",
    "end\n",
    "\n",
    "class EmblEntry\n",
    "  attr_accessor :entry_sequence, :chromosomal_coordinates, :source\n",
    "\n",
    "  def initialize(entry, start_entry, end_entry)\n",
    "    @entry_sequence = entry.to_biosequence\n",
    "    @chromosomal_coordinates = [start_entry, end_entry]\n",
    "    @source = \"\"\n",
    "  end\n",
    "\n",
    "  def add_chromosomal_coordinates\n",
    "    f1 = Bio::Feature.new('chromosomal_coordinates', @entry_sequence.accession)\n",
    "    f1.append(Bio::Feature::Qualifier.new('start', @chromosomal_coordinates[0]))\n",
    "    f1.append(Bio::Feature::Qualifier.new('end', @chromosomal_coordinates[1]))\n",
    "    @entry_sequence.features << f1\n",
    "  end\n",
    "\n",
    "  def add_gene_from_list(gene_name)\n",
    "    return if @entry_sequence.features.any? { |feature| feature.feature == 'gene_from_list' }\n",
    "    f1 = Bio::Feature.new('gene_from_list', gene_name)\n",
    "    @entry_sequence.features << f1\n",
    "  end\n",
    "\n",
    "  def process_features\n",
    "    add_chromosomal_coordinates\n",
    "\n",
    "    @entry_sequence.features.each do |feature|\n",
    "      featuretype = feature.feature\n",
    "      position = feature.position\n",
    "      qual = feature.assoc\n",
    "\n",
    "      if featuretype == 'source'\n",
    "        @source = qual['db_xref']\n",
    "      end\n",
    "\n",
    "      if featuretype == 'gene' && position.match(/^(complement\\()?(\\d+\\.\\.\\d+)(\\))?$/)\n",
    "        add_gene_from_list(qual['gene'])\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def process_cttctt_repeats(coordinates_to_use)\n",
    "    regex_positive = Regexp.new(search_positive.to_re)\n",
    "    regex_complementary = Regexp.new(search_complementary.to_re)\n",
    "    \n",
    "    @entry_sequence.features.each do |feature|\n",
    "      next unless feature.feature == 'exon' && feature.position.match?(/^(complement\\()?(\\d+\\.\\.\\d+)(\\))?$/)\n",
    "\n",
    "      start_exon, end_exon = get_coordinates(feature.position)\n",
    "      exon_sequence = @entry_sequence.subseq(start_exon, end_exon)\n",
    "\n",
    "      next if exon_sequence.nil? || exon_sequence.empty?\n",
    "\n",
    "      if feature.position.include?('complement')\n",
    "        regex = regex_complementary\n",
    "        strand = '-'\n",
    "        coordinates_format = \"complement(%s)\"\n",
    "      else\n",
    "        regex = regex_positive\n",
    "        strand = '+'\n",
    "        coordinates_format = \"%s\"\n",
    "      end\n",
    "\n",
    "      matches_motifs = find_motifs(exon_sequence, regex)\n",
    "      next if matches_motifs.empty? || matches_motifs.nil?\n",
    "\n",
    "      matches_motifs.each do |match|\n",
    "        start_match = match[:start]\n",
    "        end_match = match[:end]\n",
    "\n",
    "        start_motif = start_exon + start_match\n",
    "        end_motif = start_exon + end_match - 1\n",
    "\n",
    "        if coordinates_to_use == 'chromosomal coordinates'\n",
    "          start_motif, end_motif = convert_to_chromosomal_coordinates(start_motif, end_motif)\n",
    "        end\n",
    "\n",
    "        coordinates = coordinates_format % \"#{start_motif}..#{end_motif}\"\n",
    "\n",
    "        add_cttctt_repeat(coordinates, start_motif, end_motif, strand, @source)\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "\n",
    "  def add_cttctt_repeat(coordinates, start_motif, end_motif, strand, source)\n",
    "    return if @entry_sequence.features.any? { |f| f.feature == 'cttctt_repeat' && f.position == coordinates }\n",
    "\n",
    "    f1 = Bio::Feature.new('cttctt_repeat', coordinates)\n",
    "    f1.append(Bio::Feature::Qualifier.new('start', start_motif))\n",
    "    f1.append(Bio::Feature::Qualifier.new('end', end_motif))\n",
    "    f1.append(Bio::Feature::Qualifier.new('strand', strand))\n",
    "    f1.append(Bio::Feature::Qualifier.new('source', source))\n",
    "    f1.append(Bio::Feature::Qualifier.new('SO_Name', 'repeat_region'))\n",
    "    @entry_sequence.features << f1\n",
    "  end\n",
    "\n",
    "  def convert_to_chromosomal_coordinates(start_motif, end_motif)\n",
    "    chr_coordinates = @entry_sequence.features.find { |f| f.feature == 'chromosomal_coordinates' }\n",
    "    qual_chr = chr_coordinates.assoc\n",
    "    start_chr = qual_chr['start']\n",
    "    end_chr = qual_chr['end']\n",
    "\n",
    "    length_motif = end_motif - start_motif\n",
    "    start_motif = start_motif + start_chr - 1\n",
    "    end_motif = start_motif + length_motif - 1\n",
    "\n",
    "    [start_motif, end_motif]\n",
    "  end\n",
    "end\n",
    "\n",
    "def write_gff(gff, coordinates_to_use)\n",
    "  file = coordinates_to_use == 'chromosomal coordinates' ? 'AT_repeats_chromosomal.gff' : 'AT_repeats_sequence.gff'\n",
    "\n",
    "  File.open(file, 'w') do |file|\n",
    "    file.puts gff.to_s\n",
    "  end\n",
    "end\n",
    "\n",
    "#-------------------------------------------MAIN CODE ------------------------------------------------------------------\n",
    "\n",
    "# BEFORE EVERYTHING, SEARCHING EMBL FILE (Task 1)\n",
    "# ...\n",
    "# FIRST PART WITH SEQUENCE COORDINATES (Tasks 2, 3, 4a, 4b)\n",
    "processor = EmblProcessor.new('AT_sequences.embl')\n",
    "gff = processor.load_to_gff('sequences coordinates')\n",
    "write_gff(gff, 'sequences coordinates')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "\n",
    "# Replace 'your_embl_file.embl' with the path to your EMBL file\n",
    "embl_file_path = 'your_embl_file.embl'\n",
    "\n",
    "# Create a new GFF3 file (replace 'output.gff3' with your desired output file name)\n",
    "gff3_file_path = 'output.gff3'\n",
    "gff3_file = File.open(gff3_file_path, 'w')\n",
    "\n",
    "# Read the EMBL file\n",
    "embl_entries = Bio::FlatFile.auto(embl_file_path)\n",
    "\n",
    "# Iterate through each entry in the EMBL file\n",
    "embl_entries.each do |entry|\n",
    "  # Extract information from the EMBL entry to create GFF3 entries\n",
    "  entry.features.each do |feature|\n",
    "    # Customize this section based on your EMBL file structure and desired GFF3 format\n",
    "    seqid   = entry.definition\n",
    "    source  = 'YourSource'  # Provide a suitable source name\n",
    "    type    = feature.feature\n",
    "    start   = feature.position.from\n",
    "    end_pos = feature.position.to\n",
    "    score   = '.'  # Replace with the appropriate score if available\n",
    "    strand  = feature.position.strand\n",
    "    phase   = '.'  # Replace with the appropriate phase if available\n",
    "\n",
    "    # Build the GFF3 line\n",
    "    gff3_line = \"#{seqid}\\t#{source}\\t#{type}\\t#{start}\\t#{end_pos}\\t#{score}\\t#{strand}\\t#{phase}\\t\"\n",
    "\n",
    "    # Add attributes based on the EMBL feature qualifiers\n",
    "    attributes = feature.qualifiers.map { |qualifier| \"#{qualifier.qualifier}=#{qualifier.value}\" }\n",
    "    gff3_line += attributes.join(';')\n",
    "\n",
    "    # Write the GFF3 line to the output file\n",
    "    gff3_file.puts(gff3_line)\n",
    "  end\n",
    "end\n",
    "\n",
    "# Close the GFF3 file\n",
    "gff3_file.close"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 3.0.0",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "3.0.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
